File: ecotrack-android/app/src/main/kotlin/com/ecotrack/domain/CarbonCalculator.kt
package com.ecotrack.domain

// ---------------------------------------------------------------------------
// EcoTrack - Carbon Calculator (Domain Layer)
// ---------------------------------------------------------------------------
// PROSE methodology:
//   P - Progressive: sealed hierarchy models all action categories
//   R - Reduced Scope: pure domain logic, no Android/IO dependencies
//   O - Orchestrated: single entry-point objects with explicit contracts
//   S - Safety: value objects enforce invariants at construction time
//   E - Explicit: KDoc on every public type and function
// ---------------------------------------------------------------------------

import kotlin.math.abs

// ---------------------------------------------------------------------------
// Action category hierarchy
// ---------------------------------------------------------------------------

/**
 * Sealed hierarchy representing the category of an eco-friendly action.
 *
 * Each sub-type maps to an ADEME 2024 emission-factor domain.
 * Using a sealed class (rather than an enum) keeps the door open for
 * category-specific data fields in future iterations without breaking
 * existing call sites.
 */
sealed class ActionCategory {

    /** Human-powered mobility — bicycle, walking, etc. */
    object Transport : ActionCategory()

    /** Food and dietary choices. */
    object Food : ActionCategory()

    /** Home or office energy consumption. */
    object Energy : ActionCategory()

    /** Purchasing and consumption choices. */
    object Consumption : ActionCategory()

    /** Waste reduction, reuse, and recycling actions. */
    object Waste : ActionCategory()
}

// ---------------------------------------------------------------------------
// Value objects
// ---------------------------------------------------------------------------

/**
 * Value object representing a carbon footprint delta in kg CO₂ equivalent.
 *
 * A **negative** value means CO₂ was avoided (the action is beneficial).
 * A **positive** value means CO₂ was emitted.
 *
 * Invariant: [kgCO2e] must be finite and not NaN.
 *
 * @property kgCO2e The signed delta expressed in kilograms of CO₂ equivalent.
 * @throws IllegalArgumentException if [kgCO2e] is NaN or infinite.
 */
@JvmInline
value class CarbonDelta(val kgCO2e: Double) {
    init {
        require(kgCO2e.isFinite()) {
            "CarbonDelta must be finite and not NaN, got: $kgCO2e"
        }
    }
}

/**
 * Value object representing a user's annual carbon footprint baseline.
 *
 * Used as a reference to express a [CarbonDelta] as a percentage of the
 * user's total yearly emissions.
 *
 * Invariant: [tCO2ePerYear] must be strictly positive and finite.
 *
 * @property tCO2ePerYear Baseline expressed in **tonnes** of CO₂ equivalent per year.
 * @throws IllegalArgumentException if [tCO2ePerYear] is not strictly positive or not finite.
 */
@JvmInline
value class FootprintBaseline(val tCO2ePerYear: Double) {
    init {
        require(tCO2ePerYear.isFinite() && tCO2ePerYear > 0.0) {
            "FootprintBaseline must be strictly positive, got: $tCO2ePerYear"
        }
    }
}

// ---------------------------------------------------------------------------
// Input data class
// ---------------------------------------------------------------------------

/**
 * Input data required to calculate a [CarbonDelta] for a single eco-action.
 *
 * @property category   The [ActionCategory] that determines the calculation strategy.
 * @property name       A human-readable label for the action (e.g. "Cycling to work").
 * @property distanceKm Optional distance in kilometres, required for [ActionCategory.Transport]
 *                      calculations. Ignored for other categories.
 */
data class CarbonInput(
    val category: ActionCategory,
    val name: String,
    val distanceKm: Double? = null
)

// ---------------------------------------------------------------------------
// Emission factors (ADEME 2024, kgCO₂e/km)
// ---------------------------------------------------------------------------

/**
 * ADEME 2024 emission factors in kgCO₂e per kilometre for transport modes.
 *
 * These constants are package-private to avoid leaking implementation details
 * while remaining accessible for testing within the same package.
 */
internal object EmissionFactors {
    /** Human-powered transport — zero direct emissions. */
    const val CYCLING: Double = 0.0

    /**
     * Average passenger car (EU mix, petrol/diesel).
     * Used as the **baseline** reference for transport delta calculations.
     */
    const val CAR: Double = 0.15

    /** Urban/regional bus (average occupancy, EU mix). */
    const val BUS: Double = 0.04

    /** Rail (long-distance, EU average electricity mix). */
    const val TRAIN: Double = 0.03

    /** Commercial aviation (economy class, per passenger-km). */
    const val FLIGHT: Double = 0.255
}

// ---------------------------------------------------------------------------
// CarbonCalculator
// ---------------------------------------------------------------------------

/**
 * Pure-function calculator that converts a [CarbonInput] into a [CarbonDelta].
 *
 * **Transport logic** — the delta is the avoided emission compared to driving
 * the same distance by car:
 * ```
 * delta = (modeFactor - CAR_FACTOR) * distanceKm
 * ```
 * A negative result means CO₂ was avoided.
 *
 * **Other categories** — for Food, Consumption, Waste, and Energy the delta
 * defaults to 0.0 unless a specialised sub-calculator is added. This is an
 * intentional YAGNI decision: the model is ready to extend without breaking
 * existing behaviour.
 *
 * No side effects. No I/O. No Android dependency.
 */
object CarbonCalculator {

    /**
     * Calculates the carbon delta for the given [input].
     *
     * @param input A fully-formed [CarbonInput] describing the action.
     * @return A [CarbonDelta] representing the CO₂ impact relative to a
     *         car-travel baseline (for transport) or zero (for other categories).
     */
    fun calculate(input: CarbonInput): CarbonDelta {
        return when (input.category) {
            is ActionCategory.Transport -> calculateTransport(input)
            is ActionCategory.Energy    -> CarbonDelta(0.0)
            is ActionCategory.Food      -> CarbonDelta(0.0)
            is ActionCategory.Consumption -> CarbonDelta(0.0)
            is ActionCategory.Waste     -> CarbonDelta(0.0)
        }
    }

    /**
     * Derives the emission factor for a transport action from its [CarbonInput.name].
     *
     * The name is matched case-insensitively against known transport mode keywords.
     * Unrecognised names fall back to the car factor (delta = 0.0).
     *
     * Supported keywords: `cycling`, `bike`, `bicycle`, `walking`, `walk`,
     * `bus`, `train`, `rail`, `flight`, `plane`, `airplane`.
     */
    private fun emissionFactorForTransport(name: String): Double {
        val lower = name.lowercase()
        return when {
            lower.contains("cycling") || lower.contains("bike") ||
            lower.contains("bicycle") || lower.contains("walking") ||
            lower.contains("walk")    -> EmissionFactors.CYCLING

            lower.contains("bus")     -> EmissionFactors.BUS

            lower.contains("train") || lower.contains("rail") -> EmissionFactors.TRAIN

            lower.contains("flight") || lower.contains("plane") ||
            lower.contains("airplane") -> EmissionFactors.FLIGHT

            else                       -> EmissionFactors.CAR
        }
    }

    private fun calculateTransport(input: CarbonInput): CarbonDelta {
        val distance = input.distanceKm ?: 0.0
        val factor = emissionFactorForTransport(input.name)
        val deltaKg = (factor - EmissionFactors.CAR) * distance
        return CarbonDelta(deltaKg)
    }
}

// ---------------------------------------------------------------------------
// FootprintCalculator
// ---------------------------------------------------------------------------

/**
 * Pure-function calculator that expresses a [CarbonDelta] as a percentage of
 * a user's annual carbon [FootprintBaseline].
 *
 * No side effects. No I/O. No Android dependency.
 */
object FootprintCalculator {

    /**
     * Returns the fraction of the annual [baseline] represented by [delta].
     *
     * The result is signed: a negative percentage means the action reduced
     * the footprint; positive means it increased it.
     *
     * Division-by-zero is structurally prevented because [FootprintBaseline]
     * enforces a strictly-positive value at construction time.  This guard
     * is kept explicitly for clarity and to satisfy static analysis tools.
     *
     * @param delta    The carbon delta to express as a fraction of [baseline].
     * @param baseline The annual footprint reference (must be strictly positive).
     * @return A signed ratio in the range (-∞, +∞), where -0.01 means
     *         "1% of the annual footprint was avoided".
     */
    fun percentageOf(delta: CarbonDelta, baseline: FootprintBaseline): Double {
        // Guard is redundant (FootprintBaseline invariant covers it) but kept
        // for defensive programming and readability.
        val baselineKg = baseline.tCO2ePerYear * 1_000.0 // convert tonnes → kg
        if (baselineKg == 0.0) return 0.0
        return delta.kgCO2e / baselineKg
    }
}

File: ecotrack-ios/Sources/EcoTrack/Domain/CarbonCalculator.swift
// CarbonCalculator.swift
// EcoTrack - Domain Layer
// Swift 6 / iOS 18
//
// PROSE: Pure function — no side effects, no UIKit/Foundation imports.
// All types are Sendable and Swift 6 compatible.
//
// Emission factors source: ADEME 2024 (kgCO2e/km)

// MARK: - ActionCategory

/// Categories of eco-friendly actions tracked by EcoTrack.
///
/// Used by `CarbonInput` to route the calculation logic in `CarbonCalculator`.
public enum ActionCategory: String, CaseIterable, Sendable {
    /// Transport mode actions (cycling, car, bus, train, flight, etc.)
    case transport
    /// Food-related actions (plant-based meals, reduced meat, etc.)
    case food
    /// Energy-related actions (renewable energy, reducing consumption, etc.)
    case energy
    /// Consumption-related actions (reusable items, second-hand goods, etc.)
    case consumption
    /// Waste-related actions (recycling, composting, etc.)
    case waste
}

// MARK: - CarbonDelta

/// Immutable value object representing the CO₂ equivalent delta produced by an action.
///
/// Negative values indicate CO₂ avoided (the action is beneficial). Positive values indicate CO₂ emitted.
/// Example: cycling instead of driving over 12 km produces −1.8 kgCO2e (a saving).
///
/// **Invariants:**
/// - `kgCO2e` is always finite (never `NaN`, never infinite).
public struct CarbonDelta: Hashable, Comparable, Sendable {

    /// CO₂ equivalent delta in kilograms.
    public let kgCO2e: Double

    /// Creates a `CarbonDelta`.
    /// - Parameter kgCO2e: CO₂ equivalent in kg. Must be finite and not NaN.
    /// - Precondition: `kgCO2e.isFinite` — crashes in debug if violated.
    public init(kgCO2e: Double) {
        precondition(kgCO2e.isFinite, "CarbonDelta.kgCO2e must be finite and never NaN")
        self.kgCO2e = kgCO2e
    }

    /// A delta of zero — no carbon saved or emitted.
    public static let zero = CarbonDelta(kgCO2e: 0.0)

    public static func < (lhs: CarbonDelta, rhs: CarbonDelta) -> Bool {
        lhs.kgCO2e < rhs.kgCO2e
    }
}

// MARK: - FootprintBaseline

/// Immutable value object representing a user's annual carbon footprint baseline.
///
/// Used by `FootprintCalculator` to express a `CarbonDelta` as a percentage of
/// the user's yearly footprint.
///
/// **Invariants:**
/// - `tCO2ePerYear` is strictly positive (> 0, never zero, never negative).
public struct FootprintBaseline: Hashable, Sendable {

    /// Annual carbon footprint in tonnes CO₂ equivalent.
    public let tCO2ePerYear: Double

    /// Creates a `FootprintBaseline`.
    /// - Parameter tCO2ePerYear: Annual footprint in tCO2e. Must be > 0.
    /// - Throws: `CarbonCalculatorError.invalidCategory` if the value is ≤ 0 or not finite.
    public init(tCO2ePerYear: Double) throws {
        guard tCO2ePerYear.isFinite, tCO2ePerYear > 0 else {
            throw CarbonCalculatorError.invalidBaseline
        }
        self.tCO2ePerYear = tCO2ePerYear
    }
}

// MARK: - CarbonInput

/// Value object carrying all the data needed to compute a `CarbonDelta`.
///
/// - `category`: Determines the calculation branch.
/// - `name`: Human-readable label for the action.
/// - `distanceKm`: Required for transport actions; optional for other categories.
public struct CarbonInput: Hashable, Sendable {

    /// The category of the eco action.
    public let category: ActionCategory

    /// Human-readable name for the action (e.g. "Cycling commute").
    public let name: String

    /// Distance in kilometres, required for transport actions.
    public let distanceKm: Double?

    /// Creates a `CarbonInput`.
    /// - Parameters:
    ///   - category: The action category.
    ///   - name: The action name.
    ///   - distanceKm: Distance in km (required for `.transport`).
    public init(category: ActionCategory, name: String, distanceKm: Double? = nil) {
        self.category = category
        self.name = name
        self.distanceKm = distanceKm
    }
}

// MARK: - CarbonCalculator

/// Pure namespace for carbon delta calculation.
///
/// `CarbonCalculator` is a caseless enum used as a namespace — it cannot be
/// instantiated. All logic lives in its static functions.
///
/// ## Transport calculation
///
/// The delta is the difference between the chosen mode and a baseline car journey:
///
/// ```
/// delta = (emissionFactor - carFactor) * distanceKm
/// ```
///
/// Using ADEME 2024 emission factors (kgCO2e/km):
///
/// | Mode    | Factor   |
/// |---------|----------|
/// | Cycling | 0.000    |
/// | Car     | 0.150    |
/// | Bus     | 0.040    |
/// | Train   | 0.030    |
/// | Flight  | 0.255    |
///
/// ## Non-transport categories
///
/// Food, energy, consumption, and waste actions without a distance return a
/// delta of `0.0 kgCO2e` (no distance-based calculation defined yet).
public enum CarbonCalculator {

    // MARK: Emission factors (ADEME 2024, kgCO2e/km)

    private static let cyclingFactor: Double = 0.000
    private static let carFactor:     Double = 0.150   // reference baseline
    private static let busFactor:     Double = 0.040
    private static let trainFactor:   Double = 0.030
    private static let flightFactor:  Double = 0.255

    // MARK: Public API

    /// Computes the CO₂ delta for the given input compared to a car baseline.
    ///
    /// For transport actions the formula is:
    /// ```
    /// delta = (emissionFactor − carFactor) × distanceKm
    /// ```
    /// A negative result means CO₂ was avoided.
    ///
    /// Non-transport categories without a distance return `.zero`.
    ///
    /// - Parameter input: The action description.
    /// - Returns: A `CarbonDelta` (always finite).
    /// - Throws: `CarbonCalculatorError.missingDistance` when `input.category == .transport`
    ///   and `input.distanceKm` is `nil`.
    public static func calculate(_ input: CarbonInput) throws -> CarbonDelta {
        switch input.category {
        case .transport:
            return try calculateTransport(input)
        case .food, .energy, .consumption, .waste:
            // No distance-based model yet; delta defaults to zero.
            return .zero
        }
    }

    // MARK: Private helpers

    private static func calculateTransport(_ input: CarbonInput) throws -> CarbonDelta {
        guard let distanceKm = input.distanceKm else {
            throw CarbonCalculatorError.missingDistance
        }

        let factor = emissionFactor(for: input.name)
        let deltaValue = (factor - carFactor) * distanceKm
        return CarbonDelta(kgCO2e: deltaValue)
    }

    /// Maps a transport action name to its ADEME 2024 emission factor.
    ///
    /// The lookup is case-insensitive and falls back to the car factor (zero delta)
    /// when the name is unrecognised — making the function total.
    private static func emissionFactor(for name: String) -> Double {
        let lowercased = name.lowercased()
        if lowercased.contains("cycl") || lowercased.contains("bike") || lowercased.contains("bicycle") || lowercased.contains("walk") {
            return cyclingFactor
        } else if lowercased.contains("bus") || lowercased.contains("coach") {
            return busFactor
        } else if lowercased.contains("train") || lowercased.contains("rail") || lowercased.contains("metro") || lowercased.contains("subway") || lowercased.contains("tram") {
            return trainFactor
        } else if lowercased.contains("flight") || lowercased.contains("fly") || lowercased.contains("plane") || lowercased.contains("airplane") {
            return flightFactor
        } else {
            // Default: car (delta = 0)
            return carFactor
        }
    }
}

// MARK: - FootprintCalculator

/// Pure namespace for expressing a `CarbonDelta` as a fraction of an annual baseline.
///
/// Like `CarbonCalculator`, this is a caseless enum used as a namespace.
public enum FootprintCalculator {

    /// Returns the percentage of the annual baseline represented by the delta.
    ///
    /// The baseline is in tonnes CO₂e/year; the delta is in kg CO₂e.
    /// The conversion factor is 1 tonne = 1 000 kg.
    ///
    /// ```
    /// percentage = (delta.kgCO2e / (baseline.tCO2ePerYear × 1000)) × 100
    /// ```
    ///
    /// A zero delta always returns `0.0` without dividing.
    ///
    /// - Parameters:
    ///   - delta: The CO₂ delta to express as a percentage.
    ///   - baseline: The user's annual footprint baseline.
    /// - Returns: Percentage (can be negative if the action avoids carbon).
    public static func percentageOf(_ delta: CarbonDelta, baseline: FootprintBaseline) -> Double {
        guard delta.kgCO2e != 0.0 else { return 0.0 }
        let baselineKg = baseline.tCO2ePerYear * 1_000.0
        return (delta.kgCO2e / baselineKg) * 100.0
    }
}

// MARK: - CarbonCalculatorError

/// Errors thrown by `CarbonCalculator` and related types.
public enum CarbonCalculatorError: Error, Equatable {

    /// A transport calculation was attempted without a `distanceKm` value.
    case missingDistance

    /// The action category does not support the requested calculation.
    case invalidCategory

    /// A `FootprintBaseline` was initialised with a non-positive value.
    case invalidBaseline
}
